core/src/main/java/com/oath/oak/Chunk.java

```diff
@@ -569,12 +664,15 @@ boolean pointToValue(OpData opData) {
      * used by put/putIfAbsent/remove and rebalancer
      */
     private boolean pointToValueCAS(OpData opData) {
-        if (casEntriesArray(opData.entryIndex, OFFSET_HANDLE_INDEX, opData.prevHandleIndex, opData.handleIndex)) {
+        if (casEntriesArrayLong(opData.entryIndex, OFFSET.VALUE_REFERENCE, opData.oldValueReference,
+                opData.newValueReference)) {
             // update statistics only by thread that CASed
-            if (opData.prevHandleIndex < 0 && opData.handleIndex > 0) { // previously a remove
+            int oldValueBlock = UnsafeUtils.longToInts(opData.oldValueReference)[BLOCK_ID_LENGTH_ARRAY_INDEX];
+            int newValueBlock = UnsafeUtils.longToInts(opData.newValueReference)[BLOCK_ID_LENGTH_ARRAY_INDEX];
+            if (oldValueBlock == INVALID_BLOCK_ID && newValueBlock != INVALID_BLOCK_ID) { // previously a remove
                 statistics.incrementAddedCount();
                 externalSize.incrementAndGet();
-            } else if (opData.prevHandleIndex > 0 && opData.handleIndex == -1) { // removing
+            } else if (oldValueBlock != INVALID_BLOCK_ID && newValueBlock == INVALID_BLOCK_ID) { // removing
```

Comment:
What is the order between "block ID set to invalid" and "vale reference set to DELETED"?